[Windows] Restrict searchpath of dbghelp.dll to System32

jofrn opened the pull request:
LoadLibraryW will lookup dlls in user directories if its search path is left unrestricted. This is a security vulnerability as one can name a shared library the same as that of a system dll in order to run arbitrary code when the shared library is loaded from the path in a user directory. This change modifies it to only search within sys32 when loading dbghelp.dll.
  - - -
We can either supply an absolute path or restrict the search path itself:
1. ~~This fix ensures a secure path is found during buildtime so as to reference the absolute path when calling LoadLibraryW.~~ (must consider cross-compilation when finding the absolute path...)
2. This fix ensures a secure path is found during runtime when calling LoadLibraryExA via restricting the search path to only search system directories.

jofrn committed:
[Windows] Find absolute path to dll

mstorsjo commented:
I don't see an explanation why this is desireable to do.

This doesn't work when cross compiling, because the files used at runtime won't be available there. And it most probably can make binaries unportable, in case the files are available in different directories on different machines.

So this is a clear -1 from me for this patch, even if the motivation would be explained.

jofrn commented:
> And it most probably can make binaries unportable, in case the files are available in different directories on different machines.
It depends on how portable find_library searching vs LoadProcessW searching compare to one another. find_library is more general and configurable. We can restrict its search path further; having it match LoadProcessW for all Windows might be tricky.

> This doesn't work when cross compiling, because the files used at runtime won't be available there. 
Will take a look... thank you.

mstorsjo commented:
> > And it most probably can make binaries unportable, in case the files are available in different directories on different machines.
> 
> It depends on how portable find_library searching vs LoadProcessW searching compare to one another. find_library is more general and configurable. We can restrict its search path further; having it match LoadProcessW for all Windows might be tricky.

My point is, the absolute path where DLLs exist can differ between machines. On some versions of Windows, the default is `c:\winnt`, on others it's `c:\windows`. Perhaps it's even configurable at install time? So if you build your binaries on one machine, they won't be able to load these DLLs on one with a different layout.

Worse - if you have a i386 binary, running on i386 Windows, it will perhaps find these DLLs in `c:\windows\system32`. However on an x86_64 Windows install, there are DLLs with those names in that directory, but they're incompatible; the right ones will be in `c:\windows\SysWOW64`.

jofrn commented:
Your comments are valuable. Ensuring it can cross-compile through this route may be tricky; it looks like it is feasible to modify the search path of LoadLibraryW directly, so that may prove to be a better solution.

jofrn committed:
[Windows] Restrict search path of LoadLibraryW.

LoadLibraryW search will be more portable than searching during buildtime. This modifies its search set to omit the current directory.

dpaoliello commented on llvm/lib/Support/InitLLVM.cpp:
```diff
@@ -43,6 +43,10 @@ InitLLVM::InitLLVM(int &Argc, const char **&Argv,
   Initialized = true;
 #endif
+#ifdef _WIN32
+  // Avoid searching the directory from which the application is loaded.
```
That's not what `SetDllDirectoryA("")` does, from <https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setdlldirectorya#parameters>:
"If this parameter is an empty string (""), the call removes the current directory from the default DLL search order."

So, this is removing the current directory (not the directory where the application is loaded) from the search order: that's a good thing.

Except that Windows already prefers system and the application directory over the current directory by default: <https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order#standard-search-order-for-unpackaged-apps>

If removing the current directory from the search path is the desired behavior, then I'd recommend using `SetDefaultDllDirectories(LOAD_LIBRARY_SEARCH_DEFAULT_DIRS)` instead, since it's much easier to understand.

jofrn replied:
The intention is to remove the folder from which the application is loaded, not the current directory, as per the comment; thank you for catching this!

dpaoliello replied:
Loading dlls from the directory where the application is loaded is not a security vulnerability: the attacker hasn't gained anything.

If the attacker has the ability to write to the directory where the application is loaded then they don't have to play games with dropping binaries with names that collide with the system - they can modify the application's binary instead.

Furthermore, changing the behavior of `LoadLibrary` doesn't help with the system libraries that are loaded statically, for that you need the `/DEPENDENTLOADFLAG` linker flag.

dpaoliello replied:
For reference: <https://devblogs.microsoft.com/oldnewthing/20121031-00/?p=6203>
"Remember, the directory is the application bundle. If you drop your application into a random directory, you’ve just added everything in that directory to your bundle. And if you don’t secure your application directory, you’re allowing anybody to add components to your bundle."

dpaoliello replied:
And official MSRC guidance <https://msrc.microsoft.com/blog/2018/04/triaging-a-dll-planting-vulnerability/>:
"A DLL planting issue that falls into this category, Application Directory DLL planting, is treated as Defense-in-Depth issue that will be considered for updates in future versions only. We resolve any MSRC case that fall in this category as vNext consideration, mainly due to the amount of social engineering involved in the attack and the by design nature of the bug. A victim would have to be tricked into placing the malicious DLL (malware) where it can be triggered AND perform a non-recommended action (like running an installer in the same directory as the malware). A non-installed application has no reference point for “known good directory/binaries”, unless it creates the directory itself. Ideally, the installer should create a temporary directory with a randomized name (to prevent further DLL planting), extract its binaries to it and use them to install the application. While an attacker can make use of a drive-by download to place the malware on the victim’s system, such as into the “Downloads” folder, the essence of the attack is social engineering."

jofrn replied:
> Loading dlls from the directory where the application is loaded is not a security vulnerability: the attacker hasn't gained anything.
> 
> If the attacker has the ability to write to the directory where the application is loaded then they don't have to play games with dropping binaries with names that collide with the system - they can modify the application's binary instead.

Although that is true when we are dealing with unsigned binaries, it does not when dealing with signed ones. After modifying the binary directly, the attacker won't be able to reauthenticate it. The renamed, fake dll remains a vulnerability because it does not have to be signed in order to be loaded, so we have to ensure the search path does not find these system libraries where a user can place them.

dpaoliello replied:
> Although that is true when we are dealing with unsigned binaries, it does not when dealing with signed ones. After modifying the binary directly, the attacker won't be able to reauthenticate it. The renamed, fake dll remains a vulnerability because it does not have to be signed in order to be loaded, so we have to ensure the search path does not find these system libraries where a user can place them.

Unless you are using cert-pinning or allow-listing, then the signature doesn't matter: the attacker can replace the application with a correctly signed malicious application.

jofrn replied:
It may be the case that one has cert-pinning or allow-listing so as to be safe. And if the search path is restricted, then one can be safer.

jofrn replied:
> If removing the current directory from the search path is the desired behavior, then I'd recommend using `SetDefaultDllDirectories(LOAD_LIBRARY_SEARCH_DEFAULT_DIRS)` instead, since it's much easier to understand.

Reworked the commit to instead use LoadLibraryExA(..., LOAD_LIBRARY_SEARCH_SYSTEM32). It seems LOAD_LIBRARY_SEARCH_DEFAULT_DIRS will search user directories, among other flags listed, which we should avoid since falling back to finding dbghelp.dll or dbgcore.dll here is still dangerous. There doesn't seem to be a flag to omit just the load directory.

jofrn replied:
Oh, and thanks a lot for the information. All of these comments were very helpful.

jofrn committed:
[Windows] Restrict app load directory from dll search path (in addition to pwd)

arsenm commented on llvm/lib/Support/InitLLVM.cpp:
```diff
@@ -44,8 +44,24 @@ InitLLVM::InitLLVM(int &Argc, const char **&Argv,
+  (void)SetCurrentDirectory(LoadPath);
+  (void)SetDllDirectoryA("");
+  (void)SetCurrentDirectory(CurrentPath); // reset cwd
+  // TODO: check for errors
```
Always check for errors, this can't be left as todo

jofrn replied:
Ok, need to ensure my current approach is proper too.

jofrn committed:
[Windows] Use LoadLibraryExA and SEARCH_SYSTEM32

jofrn committed:
[Windows] Forwarded function is resolved to sys32

This was only required when referencing the shared libraries with absolute paths. The forwarded function no longer needs to be looked up separately since restricting the search path forwards functions properly.

dpaoliello reviewed (approved):
Just to be clear, this means that we will never load the version of Dbghelp.dll that ships with the [Debugging Tools for Windows](https://learn.microsoft.com/en-us/windows/win32/debug/dbghelp-versions). I don't think that it will be an issue for `MiniDumpWriteDump`, so I'm approving this.
